### PR TITLE
Add code review guideline

### PR DESCRIPTION
**Pull request by lukpueh**
Add guideline document about code reviews that may be used in addition to the development guideline document by code authors and code reviewers.

This document is largely based on a proposal by @aaaaalbert (see #17). It aims at making his proposal more succinct and adds post-review code author guidelines. Thanks, Albert!

I suggest @jhdalek55 and/or @JustinCappos as reviewers. Please review for conciseness and unnecessary redundancy with our [development guidelines](https://github.com/secure-systems-lab/lab-guidelines/blob/master/dev-workflow.md). Feel free to edit on this branch.

### CODE REVIEW

**Comment by lukpueh** [comment]
Note that the current version of the document does not include review examples, as suggested in #17. Although those might be helpful, I think the document is also valuable as is.

**Comment by jhdalek55** [comment]
I will try to take a look later today.

**Review by aaaaalbert — COMMENTED** [review]
Thanks for digging this up (and attributing it 😄)!

Here's what I did to review your commit:
* Read over it
* Compare it with my old proposal in #17

File: code-review.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+Code reviews are important, both to ensure code quality, and to learn from each
+other. To get the most out of it, code authors should voice their expectations
+of a review, and reviewers should state the performed actions and areas of
+focus.
+
+The necessary prerequisite for a *line-by-line* review is that code authors
+make sure that the changes they propose actually make sense and work at a basic
+level. *Corollary: It should never happen that neither the author nor the
+reviewer actually tests the fix.* It is still okay to request feedback on work
```
Comment (aaaaalbert) [inline code-review.md]: I would rephrase this and explicitly (albeit redundantly) request that unit / integration tests must be run and must pass. --> remove "work at a basic level", add "the unit and integration tests pass".

Also, nobody knows what a corollary is 😉.

File: code-review.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+in progress, but especially then it is important to be explicit about the depth
+of the desired review.
+
+__As code author (pre-review):__
+
+1. Make sure your proposed change does what you think it should do. This also
+   means that it conforms to the [code
+   style](https://github.com/secure-systems-lab/code-style-guidelines) etc. *Do
+   not waste the reviewer's time with things you can easily get right in the
+   first place!*
```
Comment (aaaaalbert) [inline code-review.md]: Similar to the above, add "passing tests" as a requirement?

File: code-review.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+1. Send a PR, that follows our [PR
+   guideline](commits.md#pull-request-guidelines-wip), and describe what you
+   think are the aspects that require the largest share of a reviewer's
+   attention:
+   > "Please do / look for / look at the following: ...."
+
+1. Feel free to suggest the review to someone, e.g. by using [GitHub's *review
```
Comment (aaaaalbert) [inline code-review.md]: Perhaps remove "Feel free to" so that a review is always suggest to a reviewer and doesn't end up in limbo.

File: code-review.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+   request*](https://help.github.com/en/articles/requesting-a-pull-request-review)
+   feature or by leaving a comment. Don't don't be afraid to remind them, if no
```
Comment (aaaaalbert) [inline code-review.md]: Typo: "Don't don't".

File: code-review.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+   one has been on it after a couple of days.
+
+
+__As code reviewer:__
+
+1. Inform the world about your review intent, e.g. by using [GitHub's
```
Comment (aaaaalbert) [inline code-review.md]: "Inform the world" feels quite informal. "State your intent"?

File: code-review.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+   *self-assign*](https://help.github.com/en/articles/assigning-issues-and-pull-requests-to-other-github-users)
+   feature, or by leaving a comment.
+
+1. [Perform the
+   review](https://help.github.com/en/articles/reviewing-proposed-changes-in-a-pull-request).
+   Be diligent, but also remember that the goal is not to demonstrate your
+   review skills but to integrate the proposed bug fix or feature. And be
+   respectful.
```
Comment (aaaaalbert) [inline code-review.md]: A great thing to say 👍

File: code-review.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+__As code author (post-review):__
+
+1. Address **all (!)** review comments. *Do not waste the reviewer's time by
+   having them leave the same comment twice!*
+
+    - Reply to the comment, if it is unclear, or you think the reviewer got
+      something wrong.
+
+    - Push new commits that address the comment, otherwise. These commits
```
Comment (aaaaalbert) [inline code-review.md]: Permute: "Otherwise, push new commits [...]" so that it's clear that this is another conditional branch.

**Comment by jhdalek55** [comment]
Hey all. Sorry I have not checked in on this. We were scheduled to move last week and then didn't and that plus a bunch of other open commitments has slowed me down a bit. I will take a look either later today or tomorrow. Till then, it's nice seeing Albert's name pop up in my email. Hope all is well with you, Albert.

File: code-review.md (hunk 4 above)
Comment (lukpueh) [reply code-review.md]: But I did mean, that people should be afraid. :P (thanks for catching this)

**Comment by lukpueh** [comment]
Thanks for the kind review, @aaaaalbert. I integrated your suggestions and tried to clarify a few things in 9e81129. Please, let me know what you think.

**Comment by aaaaalbert** [comment]
My pleasure ❤️ , let's do this interactively!

**Comment by lukpueh** [comment]
Thanks for the great word smithing, @jhdalek55, and for reviewing and merging, @JustinCappos! It seems that we are good to merge the entire document.
